File: frontend/js/views/reverse_mortgage.js
// Reverse Mortgage (HECM) — FHA-insured Home Equity Conversion
// Mortgage. Borrower must be ≥62, principal residence, no income/credit
// test. Lender pays YOU (lump sum, line of credit, monthly tenure, term).
// Balance compounds; due when borrower moves out, sells, or dies.
// Non-recourse: heirs never owe more than home value.
//
// HUD Principal Limit Factor (PLF) tables key off age + expected rate.
// This impl uses a simple linear approximation that captures the right
// shape (older + lower rates = more cash). Use HUD's actual PLF table
// for a binding quote.

import { api } from '../api.js';
import { esc } from '../util.js';
import { applyUiI18n, t } from '../i18n.js';

export async function renderReverseMortgage(mount, _state) {
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.reverse_mortgage.title">// REVERSE MORTGAGE (HECM)</span></h1>
        <p class="muted small" data-i18n-html="view.reverse_mortgage.intro">
            FHA-insured Home Equity Conversion Mortgage. Borrower ≥62, home
            is primary residence. Lender pays YOU; balance compounds; due
            when you leave the home permanently. Non-recourse: heirs never
            owe more than the home is worth. 2025 FHA lending limit:
            <strong>$1,209,750</strong>. PLF below is a linear approximation —
            use HUD's official table for a real quote.
        </p>
        <div class="chart-panel">
            <div style="display:grid;grid-template-columns:repeat(auto-fit,minmax(180px,1fr));gap:8px;margin-bottom:12px">
                <label>
                    <span class="muted small" data-i18n="view.reverse_mortgage.field.age">Youngest borrower age</span>
                    <input type="number" id="rm-age" step="1" min="62" max="100" value="70" style="width:100%">
                </label>
                <label>
                    <span class="muted small" data-i18n="view.reverse_mortgage.field.value">Home value $</span>
                    <input type="number" id="rm-value" step="10000" min="0" value="650000" style="width:100%">
                </label>
                <label>
                    <span class="muted small" data-i18n="view.reverse_mortgage.field.mortgage">Existing mortgage $</span>
                    <input type="number" id="rm-mortgage" step="1000" min="0" value="120000" style="width:100%">
                </label>
                <label>
                    <span class="muted small" data-i18n="view.reverse_mortgage.field.rate">Expected rate %</span>
                    <input type="number" id="rm-rate" step="0.1" min="3" max="12" value="7.5" style="width:100%">
                </label>
                <label>
                    <span class="muted small" data-i18n="view.reverse_mortgage.field.appr">Home appreciation %/yr</span>
                    <input type="number" id="rm-appr" step="0.1" min="-5" max="10" value="3.0" style="width:100%">
                </label>
                <label>
                    <span class="muted small" data-i18n="view.reverse_mortgage.field.payout">Payout type</span>
                    <select id="rm-payout" style="width:100%">
                        <option value="lump" selected data-i18n="view.reverse_mortgage.opt.lump">Lump sum</option>
                        <option value="tenure" data-i18n="view.reverse_mortgage.opt.tenure">Tenure (monthly for life)</option>
                        <option value="line" data-i18n="view.reverse_mortgage.opt.line">Line of credit</option>
                    </select>
                </label>
            </div>
            <button class="btn btn-sm primary" id="rm-run" data-i18n="view.reverse_mortgage.btn.run">⚡ Compute</button>
            <div id="rm-result" style="margin-top:12px"></div>
        </div>
    `;
    applyUiI18n(mount);
    mount.querySelectorAll('#rm-age, #rm-value, #rm-mortgage, #rm-rate, #rm-appr, #rm-payout').forEach(el => {
        el.addEventListener('input', () => compute(mount));
    });
    mount.querySelector('#rm-run').addEventListener('click', () => compute(mount));
    compute(mount);
}

async function compute(mount) {
    const result = mount.querySelector('#rm-result');
    const body = {
        age: Math.max(62, parseInt(mount.querySelector('#rm-age').value, 10) || 62),
        home_value_usd: parseFloat(mount.querySelector('#rm-value').value) || 0,
        existing_mortgage_usd: parseFloat(mount.querySelector('#rm-mortgage').value) || 0,
        expected_rate_pct: parseFloat(mount.querySelector('#rm-rate').value) || 0,
        appreciation_pct: parseFloat(mount.querySelector('#rm-appr').value) || 0,
        payout: mount.querySelector('#rm-payout').value,
    };
    try {
        const r = await api.calcReverseMortgage(body);
        renderResult(result, r, body);
    } catch (e) {
        result.innerHTML = `<p class="neg">${esc(String(e))}</p>`;
    }
}

function renderResult(result, r, body) {
    const subMaxClaim = t('view.reverse_mortgage.sub.max_claim', { home: '$' + fmt(body.home_value_usd, 0) });
    const subPlf = t('view.reverse_mortgage.sub.plf', { age: String(body.age), rate: fmt(body.expected_rate_pct, 2) + '%' });
    let payoutDescription;
    if (r.payout === 'tenure') {
        payoutDescription = t('view.reverse_mortgage.payout.tenure', {
            amount: '$' + fmt(r.monthly_payment_usd, 0), years: String(r.years_expected),
        });
    } else if (r.payout === 'line') {
        payoutDescription = t('view.reverse_mortgage.payout.line', { amount: '$' + fmt(r.net_available_usd, 0) });
    } else {
        payoutDescription = t('view.reverse_mortgage.payout.lump', { amount: '$' + fmt(r.net_available_usd, 0) });
    }
    const note = t('view.reverse_mortgage.note', {
        bg: fmt(r.balance_growth_pct, 2) + '%', appr: fmt(body.appreciation_pct, 1) + '%',
    });
    const statusIntact = t('view.reverse_mortgage.status.intact');
    const statusUnderwater = t('view.reverse_mortgage.status.underwater');
    const rows = r.rows.map((row) => `<tr>
                <td><strong>${row.year}</strong></td>
                <td class="neg">$${fmt(row.loan_balance_usd, 0)}</td>
                <td>$${fmt(row.home_value_usd, 0)}</td>
                <td class="${row.equity_remaining_usd > 0 ? 'pos' : 'neg'}">$${fmt(row.equity_remaining_usd, 0)}</td>
                <td class="${row.underwater ? 'neg' : 'muted'}">${esc(row.underwater ? statusUnderwater : statusIntact)}</td>
            </tr>`).join('');
    result.innerHTML = `
        <div class="cards" style="display:grid;grid-template-columns:repeat(auto-fit,minmax(180px,1fr));gap:8px;margin-bottom:12px">
            <div class="card"><div class="label" data-i18n="view.reverse_mortgage.card.max_claim">Max claim amount</div><div class="value">$${fmt(r.max_claim_usd, 0)}</div><div class="muted small">${esc(subMaxClaim)}</div></div>
            <div class="card"><div class="label" data-i18n="view.reverse_mortgage.card.plf">Principal Limit Factor (PLF)</div><div class="value">${fmt(r.plf_pct, 1)}%</div><div class="muted small">${esc(subPlf)}</div></div>
            <div class="card"><div class="label" data-i18n="view.reverse_mortgage.card.principal">Principal limit</div><div class="value">$${fmt(r.principal_limit_usd, 0)}</div></div>
            <div class="card"><div class="label" data-i18n="view.reverse_mortgage.card.costs">Costs (MIP + origination + closing)</div><div class="value neg">-$${fmt(r.total_costs_usd, 0)}</div></div>
            <div class="card"><div class="label" data-i18n="view.reverse_mortgage.card.payoff">Existing mortgage payoff</div><div class="value neg">-$${fmt(body.existing_mortgage_usd, 0)}</div></div>
            <div class="card"><div class="label" data-i18n="view.reverse_mortgage.card.net">Net available</div><div class="value pos">$${fmt(r.net_available_usd, 0)}</div><div class="muted small">${esc(payoutDescription)}</div></div>
        </div>
        <h3 class="section-title" data-i18n="view.reverse_mortgage.h3.projection">Balance vs home value over time</h3>
        <p class="muted small">${esc(note)}</p>
        <table class="trades" data-table-key="rm-rows">
            <thead><tr>
                <th data-i18n="view.reverse_mortgage.th.year">Year</th>
                <th data-i18n="view.reverse_mortgage.th.balance">Loan balance</th>
                <th data-i18n="view.reverse_mortgage.th.home">Home value</th>
                <th data-i18n="view.reverse_mortgage.th.equity">Equity remaining</th>
                <th data-i18n="view.reverse_mortgage.th.status">Status</th>
            </tr></thead>
            <tbody>${rows}</tbody>
        </table>
    `;
    applyUiI18n(result);
}

function fmt(n, d) {
    if (n == null || !Number.isFinite(Number(n))) return '—';
    return Number(n).toLocaleString(undefined, { minimumFractionDigits: d, maximumFractionDigits: d });
}

File: crates/traderview-core/src/reverse_mortgage.rs
//! Reverse mortgage (HECM) estimator. An FHA-insured Home Equity Conversion
//! Mortgage pays the borrower against home equity; the balance compounds and
//! comes due when they leave the home. The cash available keys off a Principal
//! Limit Factor (PLF) — here a linear approximation of HUD's table (older
//! borrowers and lower expected rates → higher PLF) — applied to the max claim
//! (min of home value and the FHA lending limit), net of upfront MIP, the
//! origination fee, closing costs, and any existing mortgage payoff. Projects
//! the loan balance against the appreciating home value to show when a
//! max-draw loan goes "underwater" (non-recourse: heirs never owe the excess).
//! PLF is an approximation — HUD's official table governs a binding quote.
//! Faithful port of the former client-side calculator. Pure compute, not advice.

use serde::{Deserialize, Serialize};

/// FHA HECM lending limit (2025).
const FHA_MAX_2025: f64 = 1_209_750.0;

#[derive(Debug, Clone, Deserialize)]
pub struct ReverseMortgageInput {
    pub age: u32,
    pub home_value_usd: f64,
    #[serde(default)]
    pub existing_mortgage_usd: f64,
    pub expected_rate_pct: f64,
    pub appreciation_pct: f64,
    /// "lump", "tenure", or "line".
    pub payout: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct ReverseRow {
    pub year: u32,
    pub loan_balance_usd: f64,
    pub home_value_usd: f64,
    pub equity_remaining_usd: f64,
    pub underwater: bool,
}

#[derive(Debug, Clone, Serialize, PartialEq, Default)]
pub struct ReverseMortgageReport {
    pub max_claim_usd: f64,
    pub plf_pct: f64,
    pub principal_limit_usd: f64,
    pub upfront_mip_usd: f64,
    pub origination_fee_usd: f64,
    pub closing_usd: f64,
    pub total_costs_usd: f64,
    pub net_available_usd: f64,
    /// "lump", "tenure", or "line" (echoed for the view's description).
    pub payout: String,
    /// Tenure only: monthly payment for life, else 0.
    pub monthly_payment_usd: f64,
    /// Tenure only: assumed life expectancy in years (100 − age), else 0.
    pub years_expected: u32,
    pub balance_growth_pct: f64,
    pub rows: Vec<ReverseRow>,
    pub valid: bool,
}

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn round4(x: f64) -> f64 {
    (x * 10_000.0).round() / 10_000.0
}

const PROJECTION_YEARS: [u32; 6] = [1, 5, 10, 15, 20, 25];

pub fn generate(i: &ReverseMortgageInput) -> ReverseMortgageReport {
    if i.home_value_usd <= 0.0 {
        return ReverseMortgageReport::default();
    }
    let age = i.age.max(62);
    let rate = i.expected_rate_pct / 100.0;
    let appr = i.appreciation_pct / 100.0;
    let payout = i.payout.trim().to_ascii_lowercase();

    let max_claim = i.home_value_usd.min(FHA_MAX_2025);
    // Linear PLF approximation: 40% at 62, +1.2pp/yr, less 4pp per point of
    // expected rate above 5%, clamped to [20%, 75%].
    let mut plf = 0.40 + (age.saturating_sub(62) as f64) * 0.012;
    plf -= (rate - 0.05).max(0.0) * 4.0;
    plf = plf.clamp(0.20, 0.75);
    let principal_limit = max_claim * plf;

    let upfront_mip = max_claim * 0.02;
    let orig_fee = 6000.0_f64.min(
        2500.0_f64.max(max_claim.min(200_000.0) * 0.02 + (max_claim - 200_000.0).max(0.0) * 0.01),
    );
    let closing = 4000.0;
    let total_costs = upfront_mip + orig_fee + closing;
    let net_available = principal_limit - i.existing_mortgage_usd - total_costs;

    let mut monthly_payment = 0.0;
    let mut years_expected = 0;
    if payout == "tenure" {
        years_expected = 100u32.saturating_sub(age);
        let r_m = (rate + 0.005) / 12.0;
        let n = (years_expected * 12) as f64;
        monthly_payment = if r_m == 0.0 {
            if n > 0.0 { net_available / n } else { 0.0 }
        } else {
            net_available * r_m / (1.0 - (1.0 + r_m).powf(-n))
        };
    }

    // Max-draw balance projection vs the appreciating home.
    let balance_growth = rate + 0.005;
    let balance0 = i.existing_mortgage_usd + total_costs + if payout == "lump" { net_available } else { 0.0 };
    let mut rows = Vec::new();
    for &yr in PROJECTION_YEARS.iter() {
        let future_balance = balance0 * (1.0 + balance_growth).powi(yr as i32);
        let future_home = i.home_value_usd * (1.0 + appr).powi(yr as i32);
        rows.push(ReverseRow {
            year: yr,
            loan_balance_usd: round2(future_balance),
            home_value_usd: round2(future_home),
            equity_remaining_usd: round2((future_home - future_balance).max(0.0)),
            underwater: future_balance > future_home,
        });
    }

    ReverseMortgageReport {
        max_claim_usd: round2(max_claim),
        plf_pct: round4(plf * 100.0),
        principal_limit_usd: round2(principal_limit),
        upfront_mip_usd: round2(upfront_mip),
        origination_fee_usd: round2(orig_fee),
        closing_usd: round2(closing),
        total_costs_usd: round2(total_costs),
        net_available_usd: round2(net_available),
        payout,
        monthly_payment_usd: round2(monthly_payment),
        years_expected,
        balance_growth_pct: round4(balance_growth * 100.0),
        rows,
        valid: true,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> ReverseMortgageInput {
        ReverseMortgageInput {
            age: 70,
            home_value_usd: 650_000.0,
            existing_mortgage_usd: 120_000.0,
            expected_rate_pct: 7.5,
            appreciation_pct: 3.0,
            payout: "lump".into(),
        }
    }

    // Pins cross-checked against the JS compute() in Python.
    #[test]
    fn default_lump() {
        let d = generate(&base());
        assert!(d.valid);
        assert!(close(d.max_claim_usd, 650_000.0));
        assert!(close(d.plf_pct, 39.6));
        assert!(close(d.principal_limit_usd, 257_400.0));
        assert!(close(d.total_costs_usd, 23_000.0));
        assert!(close(d.net_available_usd, 114_400.0));
        assert!(close(d.balance_growth_pct, 8.0));
        assert_eq!(d.rows.len(), 6);
        assert!(close(d.rows[0].loan_balance_usd, 277_992.0));
        assert!(close(d.rows[0].home_value_usd, 669_500.0));
        assert!(!d.rows[0].underwater);
        // Year 20: balance overtakes the home → underwater, zero equity.
        assert!(d.rows[4].underwater);
        assert!(close(d.rows[4].equity_remaining_usd, 0.0));
    }

    #[test]
    fn tenure_payout_monthly() {
        let d = generate(&ReverseMortgageInput { payout: "tenure".into(), ..base() });
        assert_eq!(d.years_expected, 30); // 100 − 70
        assert!(close(d.monthly_payment_usd, 839.43));
        // Non-lump: the net draw is not added to the starting balance.
        assert!(close(d.rows[0].loan_balance_usd, 143_000.0 * 1.08));
    }

    #[test]
    fn higher_rate_lowers_plf() {
        let d = generate(&ReverseMortgageInput { expected_rate_pct: 10.0, ..base() });
        // 0.40 + 8×0.012 = 0.496, less (0.10−0.05)×4 = 0.20 → 0.296.
        assert!(close(d.plf_pct, 29.6));
    }

    #[test]
    fn invalid_when_home_zero() {
        let d = generate(&ReverseMortgageInput { home_value_usd: 0.0, ..base() });
        assert!(!d.valid);
    }
}
